### ISSUE TITLE
Ax breaks if sqlalchemy>=2.0 is installed

### ISSUE THREAD

**Issue by jemissik** [facebook/Ax#1432]
### Description of problem

Ax is incompatible with sqlalchemy>=2.0, but since sqlalchemy is not a required dependency, it is not pinned in Ax's requirements. sqlalchemy might be installed due to another package needed for the project, even if it's not being used by Ax. Ax breaks if sqlalchemy >=2.0 is installed. This could reasonably happen in projects that are using visualization libraries (e.g. pyviz), projects using flask, etc.

### Package info:

```
conda install pytorch torchvision -c pytorch
pip install sqlalchemy ax-platform
```

<details> <summary> conda environment </summary>

```
# Name                    Version                   Build  Channel
anyio                     3.6.2                    pypi_0    pypi
appnope                   0.1.3                    pypi_0    pypi
argon2-cffi               21.3.0                   pypi_0    pypi
argon2-cffi-bindings      21.2.0                   pypi_0    pypi
arrow                     1.2.3                    pypi_0    pypi
asttokens                 2.2.1                    pypi_0    pypi
attrs                     22.2.0                   pypi_0    pypi
ax-platform               0.2.10                   pypi_0    pypi
backcall                  0.2.0                    pypi_0    pypi
beautifulsoup4            4.11.2                   pypi_0    pypi
blas                      1.0                    openblas  
bleach                    6.0.0                    pypi_0    pypi
botorch                   0.8.0                    pypi_0    pypi
brotlipy                  0.7.0           py310h1a28f6b_1002  
bzip2                     1.0.8                h620ffc9_4  
ca-certificates           2023.01.10           hca03da5_0  
certifi                   2022.12.7       py310hca03da5_0  
cffi                      1.15.1          py310h80987f9_3  
charset-normalizer        2.0.4              pyhd3eb1b0_0  
comm                      0.1.2                    pypi_0    pypi
cryptography              38.0.4          py310h834c97f_0  
debugpy                   1.6.6                    pypi_0    pypi
decorator                 5.1.1                    pypi_0    pypi
defusedxml                0.7.1                    pypi_0    pypi
executing                 1.2.0                    pypi_0    pypi
fastjsonschema            2.16.2                   pypi_0    pypi
ffmpeg                    4.2.2                h04105a8_0  
flit-core                 3.6.0              pyhd3eb1b0_0  
fqdn                      1.5.1                    pypi_0    pypi
freetype                  2.12.1               h1192e45_0  
gettext                   0.21.0               h826f4ad_0  
giflib                    5.2.1                h80987f9_1  
gmp                       6.2.1                hc377ac9_3  
gnutls                    3.6.15               h887c41c_0  
gpytorch                  1.9.0                    pypi_0    pypi
icu                       68.1                 hc377ac9_0  
idna                      3.4             py310hca03da5_0  
ipykernel                 6.21.2                   pypi_0    pypi
ipython                   8.10.0                   pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
ipywidgets                7.7.3                    pypi_0    pypi
isoduration               20.11.0                  pypi_0    pypi
jedi                      0.18.2                   pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
joblib                    1.2.0                    pypi_0    pypi
jpeg                      9e                   h1a28f6b_0  
jsonpointer               2.3                      pypi_0    pypi
jsonschema                4.17.3                   pypi_0    pypi
jupyter-client            8.0.3                    pypi_0    pypi
jupyter-core              5.2.0                    pypi_0    pypi
jupyter-events            0.6.3                    pypi_0    pypi
jupyter-server            2.3.0                    pypi_0    pypi
jupyter-server-terminals  0.4.4                    pypi_0    pypi
jupyterlab-pygments       0.2.2                    pypi_0    pypi
jupyterlab-widgets        1.1.2                    pypi_0    pypi
lame                      3.100                h1a28f6b_0  
lcms2                     2.12                 hba8e193_0  
lerc                      3.0                  hc377ac9_0  
libcxx                    14.0.6               h848a8c0_0  
libdeflate                1.8                  h1a28f6b_5  
libffi                    3.4.2                hca03da5_6  
libgfortran               5.0.0           11_3_0_hca03da5_28  
libgfortran5              11.3.0              h009349e_28  
libiconv                  1.16                 h1a28f6b_2  
libidn2                   2.3.1                h1a28f6b_0  
libopenblas               0.3.21               h269037a_0  
libopus                   1.3                  h1a28f6b_1  
libpng                    1.6.37               hb8d0fd4_0  
libtasn1                  4.16.0               h1a28f6b_0  
libtiff                   4.5.0                h313beb8_1  
libunistring              0.9.10               h1a28f6b_0  
libvpx                    1.10.0               hc377ac9_0  
libwebp                   1.2.4                h68602c7_0  
libwebp-base              1.2.4                h1a28f6b_0  
libxml2                   2.9.14               h8c5e841_0  
linear-operator           0.2.0                    pypi_0    pypi
llvm-openmp               14.0.6               hc6e5704_0  
lz4-c                     1.9.4                h313beb8_0  
markupsafe                2.1.2                    pypi_0    pypi
matplotlib-inline         0.1.6                    pypi_0    pypi
mistune                   2.0.5                    pypi_0    pypi
multipledispatch          0.6.0                    pypi_0    pypi
nbclassic                 0.5.2                    pypi_0    pypi
nbclient                  0.7.2                    pypi_0    pypi
nbconvert                 7.2.9                    pypi_0    pypi
nbformat                  5.7.3                    pypi_0    pypi
ncurses                   6.4                  h313beb8_0  
nest-asyncio              1.5.6                    pypi_0    pypi
nettle                    3.7.3                h84b5d62_1  
notebook                  6.5.2                    pypi_0    pypi
notebook-shim             0.2.2                    pypi_0    pypi
numpy                     1.23.5          py310hb93e574_0  
numpy-base                1.23.5          py310haf87e8b_0  
openh264                  1.8.0                h98b2900_0  
openssl                   1.1.1t               h1a28f6b_0  
opt-einsum                3.3.0                    pypi_0    pypi
packaging                 23.0                     pypi_0    pypi
pandas                    1.5.3                    pypi_0    pypi
pandocfilters             1.5.0                    pypi_0    pypi
parso                     0.8.3                    pypi_0    pypi
pexpect                   4.8.0                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pillow                    9.3.0           py310h313beb8_2  
pip                       22.3.1          py310hca03da5_0  
platformdirs              3.0.0                    pypi_0    pypi
plotly                    5.13.0                   pypi_0    pypi
prometheus-client         0.16.0                   pypi_0    pypi
prompt-toolkit            3.0.36                   pypi_0    pypi
psutil                    5.9.4                    pypi_0    pypi
ptyprocess                0.7.0                    pypi_0    pypi
pure-eval                 0.2.2                    pypi_0    pypi
pycparser                 2.21               pyhd3eb1b0_0  
pygments                  2.14.0                   pypi_0    pypi
pyopenssl                 22.0.0             pyhd3eb1b0_0  
pyro-api                  0.1.2                    pypi_0    pypi
pyro-ppl                  1.8.4                    pypi_0    pypi
pyrsistent                0.19.3                   pypi_0    pypi
pysocks                   1.7.1           py310hca03da5_0  
python                    3.10.9               hc0d8a6c_0  
python-dateutil           2.8.2                    pypi_0    pypi
python-json-logger        2.0.6                    pypi_0    pypi
pytorch                   1.13.1                 py3.10_0    pytorch
pytz                      2022.7.1                 pypi_0    pypi
pyyaml                    6.0                      pypi_0    pypi
pyzmq                     25.0.0                   pypi_0    pypi
readline                  8.2                  h1a28f6b_0  
requests                  2.28.1          py310hca03da5_0  
rfc3339-validator         0.1.4                    pypi_0    pypi
rfc3986-validator         0.1.1                    pypi_0    pypi
scikit-learn              1.2.1                    pypi_0    pypi
scipy                     1.10.0                   pypi_0    pypi
send2trash                1.8.0                    pypi_0    pypi
setuptools                65.6.3          py310hca03da5_0  
six                       1.16.0                   pypi_0    pypi
sniffio                   1.3.0                    pypi_0    pypi
soupsieve                 2.4                      pypi_0    pypi
sqlalchemy                2.0.4                    pypi_0    pypi
sqlite                    3.40.1               h7a7dc30_0  
stack-data                0.6.2                    pypi_0    pypi
tenacity                  8.2.1                    pypi_0    pypi
terminado                 0.17.1                   pypi_0    pypi
threadpoolctl             3.1.0                    pypi_0    pypi
tinycss2                  1.2.1                    pypi_0    pypi
tk                        8.6.12               hb8d0fd4_0  
torchvision               0.14.1                py310_cpu    pytorch
tornado                   6.2                      pypi_0    pypi
tqdm                      4.64.1                   pypi_0    pypi
traitlets                 5.9.0                    pypi_0    pypi
typeguard                 2.13.3                   pypi_0    pypi
typing_extensions         4.4.0           py310hca03da5_0  
tzdata                    2022g                h04d1e81_0  
uri-template              1.2.0                    pypi_0    pypi
urllib3                   1.26.14         py310hca03da5_0  
wcwidth                   0.2.6                    pypi_0    pypi
webcolors                 1.12                     pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
websocket-client          1.5.1                    pypi_0    pypi
wheel                     0.38.4          py310hca03da5_0  
widgetsnbextension        3.6.2                    pypi_0    pypi
x264                      1!152.20180806       h1a28f6b_0  
xz                        5.2.10               h80987f9_1  
zlib                      1.2.13               h5a0b063_0  
zstd                      1.5.2                h8574219_0  

```
</details>

### Traceback 

Importing the scheduler causes an error 

```
from ax.service.scheduler import scheduler
```

Traceback:
``` 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/service/scheduler.py", line 58, in <module>
    from ax.service.utils.with_db_settings_base import DBSettings, WithDBSettingsBase
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/service/utils/with_db_settings_base.py", line 24, in <module>
    from ax.storage.sqa_store.db import init_engine_and_session_factory
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/storage/sqa_store/__init__.py", line 9, in <module>
    from ax.storage.sqa_store import validation
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/storage/sqa_store/validation.py", line 11, in <module>
    from ax.storage.sqa_store.reduced_state import GR_LARGE_MODEL_ATTRS
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/storage/sqa_store/reduced_state.py", line 9, in <module>
    from ax.storage.sqa_store.sqa_classes import SQAGeneratorRun
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/ax/storage/sqa_store/sqa_classes.py", line 231, in <module>
    class SQAGeneratorRun(Base):
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_api.py", line 199, in __init__
    _as_declarative(reg, cls, dict_)
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 248, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 329, in setup_mapping
    return _ClassScanMapperConfig(
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 563, in __init__
    self._scan_attributes()
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1005, in _scan_attributes
    collected_annotation = self._collect_annotation(
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/decl_base.py", line 1233, in _collect_annotation
    extracted = _extract_mapped_subtype(
  File "/Users/jmissik/miniconda3/envs/ax-test/lib/python3.10/site-packages/sqlalchemy/orm/util.py", line 2306, in _extract_mapped_subtype
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Type annotation for "SQAGeneratorRun.arms" can't be correctly interpreted for Annotated Declarative Table form.  ORM annotations should normally make use of the ``Mapped[]`` generic type, or other ORM-compatible generic type, as a container for the actual type, which indicates the intent that the attribute is mapped. Class variables that are not intended to be mapped by the ORM should use ClassVar[].  To allow Annotated Declarative to disregard legacy annotations which don't use Mapped[] to pass, set "__allow_unmapped__ = True" on the class or a superclass this class. (Background on this error at: https://sqlalche.me/e/20/zlpr)
```

**Comment by mpolson64** [comment]
Thanks for bringing this to our attention. We spotted this when SQA 2.0 came out and [have the version pinned on our main branch](https://github.com/facebook/Ax/blob/main/setup.py), but have not published a new release since this change. We're hoping to get a new release out very soon. Until then feel free to pin `SQLAlchemy==1.4.17` in the dependencies (recommended) or installing our latest version from git (less recommended).
Again I'm hoping we'll have a version out soon, but until then I will pin this Issue to improve visibility for other users.

**Comment by madeline-scyphers** [comment]
@mpolson64 I think the fix in main won't work in general because it is still only pinned as an extra requirement, so most people won't have it pinned by default, and the code only does a [try: except ModuleNotFoundError:](https://github.com/facebook/Ax/blob/5e79a229d565859fdafb8ec8a21d4f6e4ad61195/ax/service/utils/with_db_settings_base.py#L23) to see if it is installed. If the code instead did something like this:

```python
import pkg_resources

SQLALCHEMY_MAX_VERSION = "2.0.0"

try:  # We don't require SQLAlchemy by default.
    if pkg_resources.get_distribution("sqlalchemy").version < SQLALCHEMY_MAX_VERSION:
        from ax.storage.sqa_store.db import init_engine_and_session_factory
        from ax.storage.sqa_store.decoder import Decoder
        from ax.storage.sqa_store.encoder import Encoder
        from ax.storage.sqa_store.load import (
            _get_experiment_id,
            _get_generation_strategy_id,
            _load_experiment,
            _load_generation_strategy_by_experiment_name,
        )
        from ax.storage.sqa_store.save import (
            _save_experiment,
            _save_generation_strategy,
            _save_or_update_trials,
            _update_generation_strategy,
            update_properties_on_experiment,
        )
        from ax.storage.sqa_store.sqa_config import SQAConfig
        from ax.storage.sqa_store.structs import DBSettings
        from sqlalchemy.exc import OperationalError
        from sqlalchemy.orm.exc import StaleDataError

        # We retry on `OperationalError` if saving to DB.
        RETRY_EXCEPTION_TYPES = (OperationalError, StaleDataError)
    else:
        raise ValueError("SQLAlchemy package version required to be below 2.0!")
except (pkg_resources.DistributionNotFound, ValueError):  # pragma: no cover
    DBSettings = None
    Decoder = None
    Encoder = None
    SQAConfig = None
```

That would still not require people to install sqlalchemy, but if they had sqlalchemy>2.0 from a different package, than it would still not import it.

**Comment by mpolson64** [comment]
This is addressed in Ax 0.3.0 via https://github.com/facebook/Ax/pull/1447
